### ISSUE TITLE
Feature: Add Repository Search

### ISSUE THREAD

**Issue by jasonericdavis** [open-sauced/hot#91]
### Type of feature

🍕 Feature

### Current behavior

Repositories are displayed based on the Links at the top of the page

<img width="528" alt="image" src="https://user-images.githubusercontent.com/14043845/149638723-fcd70d6c-189f-4003-8ae0-9f6e538907d1.png">

These links filter the repositories based on pre-defined filters. Currently, if a repository is outside of these filters it will never be displayed on the page making it unavailable for actions such as upvoting. 

### Suggested solution

Add the ability to search for a repository so that queries outside of the filters can be found. 

This will require an update to the UI to add a search box to allow the user to type in a repository to search for. A couple of things to consider for this functionality:

- Will the search bar always be visible or only when the user wants to search?
- It would be nice if the search returns repositories while the user is typing. This should implement the concept of debouncing to limit the number of requests to the database. This [article](https://www.freecodecamp.org/news/javascript-debounce-example/) gives a good description of this concept and how to implement it. 
 

### Additional context

No additial no context

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by hokagedemehin** [comment]
.take

**Comment by jasonericdavis** [comment]
Thank you @hokagedemehin for taking on this issue. I mistakenly didn't add the *needs-triage* label when I created the issue. My suggestion is to allow @bdougie some time to provide feedback before beginning to work on this issue.

**Comment by hokagedemehin** [comment]
> Thank you @hokagedemehin for taking on this issue. I mistakenly didn't add the *needs-triage* label when I created the issue. My suggestion is to allow @bdougie some time to provide feedback before beginning to work on this issue. 

Ko problem. I will wait

**Comment by bdougie** [comment]
I have some thoughts but not at a computer at the moment.

But in short the search doesn't need to be live search by default. If the term is captured and a search button input adds that to the fetched data. That should be sufficient for now. [producthunt.com](https://producthunt.com) has a search Omni bar on their primary navbar. As far as design UI, that is would be enough as well.

**Comment by hokagedemehin** [comment]
Hi guys, so from a positioning point of view, what do you guys think of this sample 🕵🏾‍♀️🕵🏾‍♀️🕵🏾‍♀️🕵🏾‍♂️🕵🏾‍♂️🕵🏾‍♂️

## Laptop View
![Screenshot 2022-01-18 002735](https://user-images.githubusercontent.com/16826624/149847353-416e94ad-114e-4f23-b9e0-20fc238e3908.png)

## Mobile View
![Screenshot 2022-01-18 002703](https://user-images.githubusercontent.com/16826624/149847382-128d35d6-e50a-4660-a9d3-0f149e1c445a.png)

Also, I would want to suggest that we implement React Context at this point. We could fetch the data once at the main index level and allow the fetched data to be available to any of the components that need it. This should allow us avoid props drilling in the future

Will wait for your feedback 👨🏾‍🚀

**Comment by 0-vortex** [comment]
Search should sit on the right IMHO

**Comment by hokagedemehin** [comment]
> Search should sit on the right IMHO

Thanks, @0-vortex your input is always welcome, although I don't get the meaning of the initials tho.

My laptop is going for surgery today and I think it might take me like a day to install all the applications I use regularly on my laptop depending on the network strength

I will put the search bar on the right and show the screenshot soon !!! 🚀🚀🚀

**Comment by 0-vortex** [comment]
> > Search should sit on the right IMHO
> 
> Thanks, @0-vortex your input is always welcome, although I don't get the meaning of the initials tho.
> 
> My laptop is going for surgery today and I think it might take me like a day to install all the applications I use regularly on my laptop depending on the network strength
> 
> I will put the search bar on the right and show the screenshot soon !!! 🚀🚀🚀

Sounds greate, IMHO is short for "In My Humble Opinion" 🍕

**Comment by hokagedemehin** [comment]
Hello Everyone,

Glad to be back, Surgery of my laptop was successful and we are good to go,

So what do you think of the search bar on the right side, it looks sleek

@bdougie @0-vortex @jasonericdavis 

### Laptop View
![WhatsApp Image 2022-01-21 at 01 04 14](https://user-images.githubusercontent.com/16826624/150502278-604e9c77-671d-4541-84e7-42a1502eac5e.jpeg)

### Mobile View
![WhatsApp Image 2022-01-21 at 01 04 13](https://user-images.githubusercontent.com/16826624/150502344-ab7bcbcd-6bf6-43b6-b69c-6240516ac1e7.jpeg)

**Comment by mtfoley** [comment]
I like the placement - what happens to the pizza icon when you type into the field?

**Comment by hokagedemehin** [comment]
The "search hot 🍕" is a placeholder in the search input, once you start typing, it goes away 🛹

**Comment by kelvinsekx** [comment]
Personally I feel the pizza icon looks really confusing.except there is a search icon at the extreme right.

So what I meant is that there has to be a search icon sited on the extreme right. 
![Screenshot_20220121-142904~2](https://user-images.githubusercontent.com/51183064/150535306-25b8674e-4854-457d-b893-ef54fd828682.jpg)

**Comment by hokagedemehin** [comment]
I think the best solution will be to remove the pizza icon and put a search icon there instead. At least that way, one can easily identify the search bar

**Comment by bdougie** [comment]
The search icon is preferred. The copy could be "Search Open Sauced"

Also agree on the Context being needed. The placement of the search box will prevent is from being drilled down.

**Comment by hokagedemehin** [comment]
> Also agree on the Context being needed. The placement of the search box will prevent it from being drilled down. useContext can be leveraged for the SearchContext and that can leverage in the PostWrap.

Thanks @bdougie for this suggestion on how we can implement the Context API. I will like to make a suggestion on this so for everyone to comment on. Why don't we have a DataContext.js file that fetches the data, this file wraps the main app.js file and makes the fetched data available across the entire application.  This app also utilizes useState that is also passed down as a props so that the search component for example can make changes to the data and that change can reflect in the PostGrid or PostList component straight away.

so folks, what do you think??

**Comment by 0-vortex** [comment]
> > Also agree on the Context being needed. The placement of the search box will prevent it from being drilled down. useContext can be leveraged for the SearchContext and that can leverage in the PostWrap.
> 
> Thanks @bdougie for this suggestion on how we can implement the Context API. I will like to make a suggestion on this so for everyone to comment on. Why don't we have a DataContext.js file that fetches the data, this file wraps the main app.js file and makes the fetched data available across the entire application. This app also utilizes useState that is also passed down as a props so that the search component for example can make changes to the data and that change can reflect in the PostGrid or PostList component straight away.
> 
> so folks, what do you think??

Hard no on this one, we are going to use the search box to filter an API, even if that API does not exist right now, there's no way to reason that paginated data can be shared with the 2 components while also being full-text filterable. For mock purposes, sure, iterate the existing array, but innovating on the mock is completely pointless, even the code quality improvements can live somewhere else (vite, astro refactor) ...

My understanding is that searching that box will show a drop down with available repositories starred by open-sauced users, similar to (ex:) https://openbase.com/js/@open-sauced/check-engines but having more detailed information about our user activity around that repository. Making the fetching data context global is a good idea for different reasons that would first need designing (context middlewares, have a clear understanding of the APIs we need for this).

Focusing on the performance and usability is IMHO the way to go and it's looking great so far :D

**Comment by hokagedemehin** [comment]
> My understanding is that searching that box will show a drop down with available repositories starred by open-sauced users, similar to (ex:) https://openbase.com/js/@open-sauced/check-engines but having more detailed information about our user activity around that repository. Making the fetching data context global is a good idea for different reasons that would first need designing (context middlewares, have a clear understanding of the APIs we need for this).

So the search component is supposed to filter the fetched data as the user is typing in the repo name. The search bar is not making any API calls to the backend, it is only making use of the data that is already fetched.

**Comment by 0-vortex** [comment]
> > My understanding is that searching that box will show a drop down with available repositories starred by open-sauced users, similar to (ex:) https://openbase.com/js/@open-sauced/check-engines but having more detailed information about our user activity around that repository. Making the fetching data context global is a good idea for different reasons that would first need designing (context middlewares, have a clear understanding of the APIs we need for this).
> 
> So the search component is supposed to filter the fetched data as the user is typing in the repo name. The search bar is not making any API calls to the backend, it is only making use of the data that is already fetched.

In the current iteration where all the data is available, yes, however in future iterations we will not keep local storage of 7000 repositories to search using `.filter` or `.map`. Been looking at [supabase FT search](https://supabase.com/docs/guides/database/full-text-search) but not sure that is enough even, it's most likely we will use some kind of serverless function for this 🍕

**Comment by bdougie** [comment]
Adding a DataContext is a much a larger project and out of scope for this. I would even approach this to be much a much smaller task than described above. 

The SearchContext should only capture the string input from the user. The Context will be added to the fetchedRepositorie calls in the PostsWrap component. The goal here is to filter search and lean on Supabase for the data. At the moment, caching all that result in Context is not necessary. Only store what the user is searching. 

Adding a dropdown or overlay for search is also not needed and can scoped in a future issue. 

Look the opensauced repo for an example of what I implement above with with setting the [goalsId](https://github.com/open-sauced/open-sauced/blob/eb1bb9030cbc0bba6d883d5ed933e99f5488ea14/src/components/RepositoryGoals.jsx#L17). 

https://github.com/open-sauced/open-sauced/search?q=LocaleContext

**Comment by bdougie** [comment]
Checking in @hokagedemehin. Are you able to PR what you have? I would love to assist in unblocking this.

**Comment by bdougie** [comment]
I played around with this and think this will need to remove Astro to make this a little more approachable. #80 

Example function for interacting with supabase
```js

export async function fetchWithSearch(orderBy = 'total_stars', limit = 25, searchText) {
  const { data: recommendations, error } = await supabase
    .from('recommendations')
    .select('repo_name, description,stars,issues, total_stars, avg_recency_score, contributors, votes')
    .textSearch('repo_name', `'${searchText}'`) // The string will need to be interpolated with the ''
    .limit(limit)
    .order(orderBy, { ascending: false });

  console.error(error);

  return recommendations;
}

**Comment by hokagedemehin** [comment]
> Checking in @hokagedemehin. Are you able to PR what you have? I would love to assist in unblocking this. 

Hi @bdougie thanks for reaching out. I have been engaged with work and family at the same time. I think we would need to convert the `index.astro` to `index.js` for the search functionality to work. Is that possible?

**Comment by bdougie** [comment]
> Hi @bdougie thanks for reaching out. I have been engaged with work and family at the same time. I think we would need to convert the index.astro to index.js for the search functionality to work. Is that possible?

The conversion has begun #140

**Comment by bdougie** [comment]
Sharing the latest mockup for search here:

![search (1)](https://user-images.githubusercontent.com/5713670/163283198-812e774b-2f73-4895-85ac-beb3d598a5b7.png)

**Comment by bdougie** [comment]
Going to have @shamimbinnur take this over.
